CoreServices/WindowServer/libinput/tools/libinput-analyze-touch-down-state.man
.TH libinput-analyze-touch-down-state "1"
.SH NAME
libinput\-analyze\-touch\-down\-state \- analyze the touch states
.SH SYNOPSIS
.B libinput analyze touch-down-state [\-\-help] [options] \fIrecording.yml\fI
.SH DESCRIPTION
.PP
The
.B "libinput analyze touch\-down\state"
tool analyzes a recording made with
.B "libinput record"
and prints "down" state of each touch. This tool aids with the detection of stuck touches.
.PP
This is a debugging tool only, its output may change at any time. Do not
rely on the output.
.SH OPTIONS
.TP 8
.B \-\-help
Print help
.TP 8
.B \-\-use-st
Use the single-touch BTN_TOOL_ bits instead of the slot state. The output
will only show the "highest" finger down at any time. For examples, where
two fingers are down, only the second slot will be marked as down.
.SH OUTPUT
An example output for a two-finger alternating sequence below.
.PP
.nf
.sf
  6.140281 +1062ms: x | x
  7.410377 +1257ms:   | x
  7.420200    +9ms:   |
 11.233108 +3812ms: x | x
 11.245721   +12ms: x | x
 11.850206  +604ms:   |
 13.827740 +1977ms: x |
 13.839723   +11ms: x |
 14.704027  +864ms: x | x
 14.716691   +12ms: x | x
.fi
.in
.SH LIBINPUT
Part of the
.B libinput(1)
suite


libs-gui/NEWS
1 NEWS
******

The currently released version of the library is '0.28.0'.

1.1 Noteworthy changes in version '0.28.0'
==========================================

This version adds support for modern XIB files and many new classes.
Plus the usual bunch of bug fixes.

   * Support loading of document XIB files.
   * Improve Key Value Binding for NSArrayController and add more
     bindings.
   * Better support for multi monitor usage and other improvement in the
     backend integration.
   * Add classes NSFontCollection, NSColorSampler, NSSpeechRecognizer,
     NSAppearance, NSPDFInfo, NSPICTImageRep, NSCIImageRep,
     NSPDFImageRep, NSPDFPanel, NSDataAsset, NSDatePicker,
     NSDatePickerCell, NSPredicateEditor, NSPredicateEditorRowTemplate,
     NSRuleEditor, NSGestureRecognizer, NSButtonTouchBarItem,
     NSCandidateListTouchBarItem, NSClickGestureRecognizer,
     NSColorPickerTouchBarItem, NSCustomTouchBarItem,
     NSGroupTouchBarItem, NSMagnificationGestureRecognizer,
     NSPanGestureRecognizer, NSPickerTouchBarItem,
     NSPopoverTouchBarItem, NSPressGestureRecognizer,
     NSRotationGestureRecognizer, NSSharingServicePickerTouchBarItem,
     NSSliderTouchBarItem, NSStepperTouchBarItem, NSTouchBarItem,
     NSTouchBar, NSTouch, NSDockTile.
   * Implement NSEPSImageRep.
   * Better encoding handling in RTF files.
   * Theming and drawing improvements.
   * Increase small font size to 10.
   * New cursor and stepper images.
   * Move NSFileWrapper to Foundation.
   * Fixed build on Debian GNU/kFreeBSD.
   * With command line argument -autolaunch YES, do not activate the
     application when -activateIgnoringOtherApps: is invoked.
   * Improvements to WindowMaker compatibility (e.g.  WMFHideApplication
     support).
   * Lowered NSFloatingWindowLevel by one to distinguish floating panels
     from menus.

1.2 Noteworthy changes in version '0.27.0'
==========================================

This version includes numerous bugfixes, compatibility improvements and
other changes accumulated over the last year.  It also enables work to
be done on integrating NSViews with a Core Animation renderer.

   * Make targetForAction safer.
   * Speed up menu updates.
   * Clean up speech tool compilation and switch to newer interface.
   * Fix bug in CUPS subclassing introduced in last release.
   * Minor improvements to typesetting.
   * Add NSIsControllerMarker.
   * Fix tracking on segmented cell.
   * Bring slider cell closer to Cocoa implementation.
   * Add ivar for Core Animation in NSView.
   * Improve border calculation on printing.
   * Lazy load app icon.
   * Better detection of removable volumes.
   * Polish translations.
   * Japanese translations.
   * Lots of bug fixes.

1.3 Noteworthy changes in version '0.26.2'
==========================================

This version is a small, but important bugfix release.

   * printing: Fix allocation of the CUPS printing classes.
   * installation: Fix the configure script.

1.4 Noteworthy changes in version '0.26.1'
==========================================

This version is released to conincide with version 1.25.1 of
gnustep-base, which contains changes required for this version of
gnustep-gui and gnustep-back.

   It includes an important workaround for users of GNUstep Objective-C
Runtime (libobjc2) and non-fragile ABI to avoid a bug in interaction
between the clang compiler and the runtime when non-fragile ABI is in
use.  Specifically, Clang and the runtime may disagree on what is the
offset of an ivar in a class's RAM. This manifested in a crash at
application startup due to misalignment of _gcontext inside NSThread.
See the mailing list discussion
(http://lists.gnu.org/archive/html/discuss-gnustep/2017-12/msg00129.html)
for more information.

   It also contains the following changes:

   * tests: Cleanup of warnings.
   * tests: Fix text system deallocation test.
   * printing: Undefine __BLOCKS__ before including cups.h, as some
     versions of the header expect that libdispatch is present and used
     if __BLOCKS__ is defined.
   * graphics context: Workaround for Clang+libobjc2+nonfragile ABI
     issue.

1.5 Noteworthy changes in version '0.26.0'
==========================================

This version was bumped due to previous binary incompatibilities between
0.25.0 and 0.25.1.  This version also includes numerous compatibility
improvements from the Summer of Code project, and a wide variety of
other fixes.  Notably, it fixes the use of cupsGetPPD() in the printing
system.

   * printing: Add an include to get deprecated function cupsGetPPD() on
     newer CUPS systems.
   * chore: Bump required base version.
   * tiff: Support for writing resolution.
   * jpeg: Save resolution information if it is different from 72 dpi.
   * save panel: Fix return type of sorting function.
   * events: Add some newer Cocoa enums and one method with dummy
     implementation.
   * speech synthesis: NSSpeechSynthesizerDelegate is now a @protocol on
     runtimes that support it.
   * pasteboard: New type identifiers.
   * translations: Some work on Polish, Russian and German translations
   * cell: Improvements to mouse tracking logic on NSCell.
   * image: If an unknown named image is unarchived with a coder or
     keyed coder, keep the name.
   * screen: Add -backingScaleFactor and return 1.0.
   * window: Return 1.0 from -backingScaleFactor.

   * compatibility: Numerous stub implementations of constants, classes
     and methods to improve source-level compatibility.
   * other bugfixes

1.6 Noteworthy changes in version '0.25.1'
==========================================

   * JPEG (saving) alpha channel fixes and size with resolution != 72
   * JPEG resolution read support
   * TIFF saving fixes
   * Improved volumes mounting and support
   * Portability improvements in volume mounting and support
   * Corrected layout of empty strings
   * Only update visible menus

1.7 Noteworthy changes in version '0.25.0'

CoreServices/WindowServer/libinput/tools/libinput-analyze-touch-down-state.py
#!/usr/bin/env python3
# -*- coding: utf-8
# vim: set expandtab shiftwidth=4:
# -*- Mode: python; coding: utf-8; indent-tabs-mode: nil -*- */
#
# Copyright © 2020 Red Hat, Inc.
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the 'Software'),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice (including the next
# paragraph) shall be included in all copies or substantial portions of the
# Software.
#
# THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.
#
#
# Prints the down/up state of each touch slot
#
# Input is a libinput record yaml file

import argparse
import enum
import sys
import yaml
import libevdev


class Slot:
    class State(enum.Enum):
        NONE = "NONE"
        BEGIN = "BEGIN"
        UPDATE = "UPDATE"
        END = "END"

    def __init__(self, index):
        self._state = Slot.State.NONE
        self.index = index
        self.used = False

    def begin(self):
        assert self.state == Slot.State.NONE
        self.state = Slot.State.BEGIN

    def end(self):
        assert self.state in (Slot.State.BEGIN, Slot.State.UPDATE)
        self.state = Slot.State.END

    def sync(self):
        if self.state == Slot.State.BEGIN:
            self.state = Slot.State.UPDATE
        elif self.state == Slot.State.END:
            self.state = Slot.State.NONE

    @property
    def state(self):
        return self._state

    @state.setter
    def state(self, newstate):
        assert newstate in Slot.State

        if newstate != Slot.State.NONE:
            self.used = True
        self._state = newstate

    @property
    def is_active(self):
        return self.state in (Slot.State.BEGIN, Slot.State.UPDATE)

    def __str__(self):
        return "+" if self.state in (Slot.State.BEGIN, Slot.State.UPDATE) else " "


def main(argv):
    parser = argparse.ArgumentParser(description="Print the state of touches over time")
    parser.add_argument(
        "--use-st", action="store_true", help="Ignore slots, use the BTN_TOOL bits"
    )
    parser.add_argument(
        "path", metavar="recording", nargs=1, help="Path to libinput-record YAML file"
    )
    args = parser.parse_args()

    yml = yaml.safe_load(open(args.path[0]))
    device = yml["devices"][0]
    absinfo = device["evdev"]["absinfo"]
    try:
        nslots = absinfo[libevdev.EV_ABS.ABS_MT_SLOT.value][1] + 1
    except KeyError:
        args.use_st = True

    tool_slot_map = {
        libevdev.EV_KEY.BTN_TOOL_FINGER: 0,
        libevdev.EV_KEY.BTN_TOOL_PEN: 0,
        libevdev.EV_KEY.BTN_TOOL_DOUBLETAP: 1,
        libevdev.EV_KEY.BTN_TOOL_TRIPLETAP: 2,
        libevdev.EV_KEY.BTN_TOOL_QUADTAP: 3,
        libevdev.EV_KEY.BTN_TOOL_QUINTTAP: 4,
    }
    if args.use_st:
        for bit in tool_slot_map:
            if bit.value in device["evdev"]["codes"][libevdev.EV_KEY.value]:
                nslots = max(nslots, tool_slot_map[bit])

    slots = [Slot(i) for i in range(0, nslots)]
    # We claim the first slots are used just to make the formatting
    # more consistent
    for i in range(min(5, len(slots))):
        slots[i].used = True

    slot = 0
    last_time = None
    last_slot_state = None
    header = "Timestamp | Rel time |     Slots     |"
    print(header)
    print("-" * len(header))

    def events():
        for event in device["events"]:
            for evdev in event["evdev"]:
                yield evdev

    for evdev in events():
        e = libevdev.InputEvent(
            code=libevdev.evbit(evdev[2], evdev[3]),
            value=evdev[4],
            sec=evdev[0],
            usec=evdev[1],
        )

        # single-touch formatting is simpler than multitouch, it'll just
        # show the highest finger down rather than the correct output.
        if args.use_st:
            if e.code in tool_slot_map:
                slot = tool_slot_map[e.code]
                s = slots[slot]
                if e.value:
                    s.begin()
                else:
                    s.end()
        else:
            if e.code == libevdev.EV_ABS.ABS_MT_SLOT:
                slot = e.value
                s = slots[slot]
                # bcm5974 cycles through slot numbers, so let's say all below
                # our current slot number was used
                for sl in slots[: slot + 1]:
                    sl.used = True
            else:
                s = slots[slot]
                if e.code == libevdev.EV_ABS.ABS_MT_TRACKING_ID:
                    if e.value == -1:
                        s.end()
                    else:
                        s.begin()
                elif e.code in (
                    libevdev.EV_ABS.ABS_MT_POSITION_X,
                    libevdev.EV_ABS.ABS_MT_POSITION_Y,
                    libevdev.EV_ABS.ABS_MT_PRESSURE,
                    libevdev.EV_ABS.ABS_MT_TOUCH_MAJOR,
                    libevdev.EV_ABS.ABS_MT_TOUCH_MINOR,
                ):
                    # If recording started after touch down
                    if s.state == Slot.State.NONE:
                        s.begin()

        if e.code == libevdev.EV_SYN.SYN_REPORT:
            current_slot_state = tuple(s.is_active for s in slots)

            if current_slot_state != last_slot_state:
                if last_time is None:
                    last_time = e.sec * 1000000 + e.usec
                    tdelta = 0
                else:
                    t = e.sec * 1000000 + e.usec
                    tdelta = int((t - last_time) / 1000) / 1000
                    last_time = t

                fmt = " | ".join([str(s) for s in slots if s.used])
                print(
                    "{:2d}.{:06d} | {:+7.3f}s | {}".format(e.sec, e.usec, tdelta, fmt)
                )

                last_slot_state = current_slot_state

            for s in slots:
                s.sync()


if __name__ == "__main__":
    main(sys.argv)
